numpy/doc/structured_arrays.py

```diff
@@ -6,231 +6,430 @@
+position we get the second structure::
 
  >>> x[1]
- (2,3.,"World")
+ ('Fido', 3, 27.0)
 
-Conveniently, one can access any field of the array by indexing using the
-string that names that field. ::
+Individual fields can be accessed and modified by indexing with the field name::
 
- >>> y = x['bar']
- >>> y
- array([ 2.,  3.], dtype=float32)
- >>> y[:] = 2*y
- >>> y
- array([ 4.,  6.], dtype=float32)
+ >>> x['age']
+ array([9, 3], dtype=int32)
+ >>> x['age'] = 5
  >>> x
- array([(1, 4.0, 'Hello'), (2, 6.0, 'World')],
-       dtype=[('foo', '>i4'), ('bar', '>f4'), ('baz', '|S10')])
+ array([('Rex', 5, 81.0), ('Fido', 5, 27.0)],
+       dtype=[('name', 'S10'), ('age', '<i4'), ('weight', '<f4')])
+
+The goal of structured arrays is to allow low-level manipulation of complex
+structured datatypes, for example for interpreting binary blobs of structured
+data. For this purpose numpy includes specialized features such as subarrays
+and nested datatypes, and allows detailed control over the memory layout of the
+structure. Structured datatypes are designed to mimic 'structs' in the C
+language, making them useful for interfacing with C code.
+
+Users wishing simply to manipulate tabular data with labelled columns are
+enouraged to also consider alternative pydata projects such as pandas, xarray
+or DataArray, which provide convenient and efficient relational operations on
+labelled multi-type data through a higher-level interface.
+
+.. _defining-structured-types:
+
+Structured Datatypes
+====================
+
+Structured datatypes can be thought of as a sequence of bytes of a certain
+length (the structure's :term:`itemsize`) which is interpreted as a collection
+of fields, where each field has a name, a datatype, and a byte offset within
+the structure. The datatype of a field may be any numpy datatype including
+other structured datatypes, and it may also be a :term:`sub-array` which
+behaves like an ndarray of a specified shape. The offsets of the fields are
+arbitrary, and fields may even overlap. These offsets are usually determined
+automatically by numpy but can also be manually specified.
+
+Structured Datatype Specification
+----------------------------------
+
+Structured datatypes may be created using the function :func:`numpy.dtype` with
+one of 3 alternative forms of specification which vary in flexibility and
+conciseness. The forms of specification are further documented in the
+:ref:`Data Type Objects <arrays.dtypes.constructing>` reference page, and in
+summary they are:
+
+1. A list of tuples, one tuple per field
+````````````````````````````````````````
+
+Each tuple has the form ``(fieldname, datatype, shape)`` where shape is
+optional. ``fieldname`` is a string, ``datatype`` may be any object convertible
+to a datatype, and shape is a tuple of integers specifying subarray shape.
+
+ >>> x = np.zeros(3, dtype=[('x', 'f4'), ('y', np.float32), ('z', 'f4', (2,2))])
+ >>> x
```

Comment:
I think it might be clearer to show these as `>>> dt = np.dtype(...)`, `>>> np.zeros(1, dtype=dt)`